datafusion/core/src/physical_optimizer/projection_pushdown.rs

```diff
@@ -916,7 +916,9 @@ fn update_expr(
                     .find_map(|(index, (projected_expr, alias))| {
                         projected_expr.as_any().downcast_ref::<Column>().and_then(
                             |projected_column| {
-                                column.name().eq(projected_column.name()).then(|| {
+                                (column.name().eq(projected_column.name())
```

Comment:
There is another similar check a bit further down in the file in `new_columns_for_join_on`

https://github.com/apache/arrow-datafusion/blob/c9049eda85683d958a39eca5fd3a382943ee7fa6/datafusion/core/src/physical_optimizer/projection_pushdown.rs#L1046

Likewise in new_indices_for_join_filter

https://github.com/apache/arrow-datafusion/blob/c9049eda85683d958a39eca5fd3a382943ee7fa6/datafusion/core/src/physical_optimizer/projection_pushdown.rs#L1132

I wonder if those checks need to extended to check the index as well? Maybe we could extract the "find matching projection column" logic as a function 🤔

Reply:
I am working on a new, more comprehensive and less error-prone projection optimizer. I plan to open a PR after our internal review. Just so you know, all these utils will be removed in the new version.

Reply:
If so, then I don't touch other places like `new_columns_for_join_on`, but only address the reported issue.

Reply:
ditto for `new_indices_for_join_filter`.

Reply:
If another rewrite is imminent, it is all the more important that we thoroughly test for this behavior

Reply:
I think for `new_columns_for_join_on` and `new_indices_for_join_filter`, you won't hit a similar issue like this. The reason is https://github.com/apache/arrow-datafusion/pull/9096#discussion_r1474725804.